ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java

```diff
@@ -526,16 +567,44 @@ private void handleAdminRequest(Request request) throws InterruptedException, IO
           responseQueueTimeHistogram = metrics.triggerCompactionResponseQueueTimeInMs;
           responseSendTimeHistogram = metrics.triggerCompactionResponseSendTimeInMs;
           responseTotalTimeHistogram = metrics.triggerCompactionRequestTotalTimeInMs;
-          ServerErrorCode error = validateRequest(adminRequest.getPartitionId(), false);
+          error = validateRequest(adminRequest.getPartitionId(), RequestOrResponseType.AdminRequest);
           if (error != ServerErrorCode.No_Error) {
             logger.error("Validating trigger compaction request failed with error {} for {}", error, adminRequest);
           } else if (!storageManager.scheduleNextForCompaction(adminRequest.getPartitionId())) {
             error = ServerErrorCode.Unknown_Error;
             logger.error("Triggering compaction failed for {}. Check if admin trigger is enabled for compaction",
                 adminRequest);
           }
-          response = new AdminResponse(adminRequest.getCorrelationId(), adminRequest.getClientId(), error);
+          break;
+        case RequestControl:
+          metrics.requestControlRequestQueueTimeInMs.update(requestQueueTime);
+          metrics.requestControlRequestRate.mark();
+          processingTimeHistogram = metrics.requestControlResponseQueueTimeInMs;
+          responseQueueTimeHistogram = metrics.requestControlResponseQueueTimeInMs;
+          responseSendTimeHistogram = metrics.requestControlResponseSendTimeInMs;
+          responseTotalTimeHistogram = metrics.requestControlRequestTotalTimeInMs;
+          RequestControlAdminRequest controlRequest = RequestControlAdminRequest.readFrom(requestStream, adminRequest);
+          RequestOrResponseType toControl = controlRequest.getRequestTypeToControl();
+          if (!requestsDisableInfo.containsKey(toControl)) {
```

Comment:
This is a very specific check that is done after the request is deserialized. `validateRequest()` does not have that information + this check is too specific to this particular request.